Quotient/e2ee/sssshandler.cpp

```diff
@@ -0,0 +1,253 @@
+// SPDX-FileCopyrightText: 2023 Tobias Fella <tobias.fella@kde.org>
+// SPDX-License-Identifier: LGPL-2.0-or-later
+
+#include "sssshandler.h"
+
+#include "cryptoutils.h"
+#include "qolmaccount.h"
+
+#include "../csapi/key_backup.h"
+
+#include "../events/event.h"
+
+#include "../database.h"
+#include "../logging_categories_p.h"
+#include "../room.h"
+
+using namespace Quotient;
+
+namespace {
+constexpr inline auto MegolmBackupKey = "m.megolm_backup.v1"_ls;
+constexpr inline auto CrossSigningMasterKey = "m.cross_signing.master"_ls;
+constexpr inline auto CrossSigningSelfSigningKey = "m.cross_signing.self_signing"_ls;
+constexpr inline auto CrossSigningUserSigningKey = "m.cross_signing.user_signing"_ls;
+}
+
+QByteArray SSSSHandler::decryptKey(event_type_t keyType, const QString& defaultKey,
+                                   const QByteArray& decryptionKey)
+{
+    Q_ASSERT(m_connection);
+    const auto& encryptedKeyObject = m_connection->accountData(keyType);
+    if (!encryptedKeyObject) {
+        qWarning() << "No account data for key" << keyType;
+        emit error(NoKeyError);
+        return {};
+    }
+    const auto& encrypted =
+        encryptedKeyObject->contentPart<QJsonObject>("encrypted"_ls).value(defaultKey).toObject();
+
+    auto hkdfResult = hkdfSha256(decryptionKey, QByteArray(32, u'\0'), keyType.data());
+    if (!hkdfResult.has_value()) {
+        qCWarning(E2EE) << "Failed to calculate HKDF for" << keyType;
+        emit error(DecryptionError);
+    }
+    const auto& keys = hkdfResult.value();
+
+    auto rawCipher = QByteArray::fromBase64(encrypted["ciphertext"_ls].toString().toLatin1());
+    auto result = hmacSha256(keys.mac(), rawCipher);
+    if (!result.has_value()) {
+        qCWarning(E2EE) << "Failed to calculate HMAC for" << keyType;
+        emit error(DecryptionError);
+    }
+    if (QString::fromLatin1(result.value().toBase64()) != encrypted["mac"_ls].toString()) {
+        qCWarning(E2EE) << "MAC mismatch for" << keyType;
+        emit error(DecryptionError);
+        return {};
+    }
+    auto decryptResult =
+        aesCtr256Decrypt(rawCipher, keys.aes(),
+                         asCBytes<AesBlockSize>(QByteArray::fromBase64(
+                             encrypted["iv"_ls].toString().toLatin1())));
+    if (!decryptResult.has_value()) {
+        qCWarning(E2EE) << "Failed to decrypt for" << keyType;
+        emit error(DecryptionError);
+    }
+    auto key = QByteArray::fromBase64(decryptResult.value());
+    m_connection->database()->storeEncrypted(keyType, key);
+    return key;
+}
+
+void SSSSHandler::unlockSSSSFromPassword(const QString& password)
+{
+    Q_ASSERT(m_connection);
+    unlockAndLoad(password.toLatin1(), true);
```

Comment:
More of an aside comment for now (seems that [we have a gap in the spec](https://github.com/matrix-org/matrix-spec/issues/1750) here) but this code implies that the passphrase is in Latin-1 encoding. I guess we could assume that UTF-8 would use with other clients, too.

Reply:
changed